diff --git a/src/Wizard/FormWizardBase.php b/src/Wizard/FormWizardBase.php
index e537ef8..901e0b7 100644
--- a/src/Wizard/FormWizardBase.php
+++ b/src/Wizard/FormWizardBase.php
@@ -176,6 +176,15 @@ abstract class FormWizardBase extends FormBase implements FormWizardInterface {
   }
 
   /**
+   * The translated text of the "Previous" button's text.
+   *
+   * @return \Drupal\Core\StringTranslation\TranslatableMarkup
+   */
+  public function getPrevOp() {
+    return $this->t('Previous');
+  }
+
+  /**
    * {@inheritdoc}
    */
   public function getNextParameters($cached_values) {
@@ -284,8 +293,15 @@ abstract class FormWizardBase extends FormBase implements FormWizardInterface {
         $this->machine_name = $cached_values['id'];
       }
       $this->getTempstore()->set($this->getMachineName(), $cached_values);
+      $next_parameters = $this->getNextParameters($cached_values);
       if (!$form_state->get('ajax')) {
-        $form_state->setRedirect($this->getRouteName(), $this->getNextParameters($cached_values));
+        $form_state->setRedirect($this->getRouteName(), $next_parameters);
+      }
+      else {
+        // Switch steps for ajax forms.
+        if (!empty($next_parameters['step'])) {
+          $this->step = $next_parameters['step'];
+        }
       }
     }
   }
@@ -309,7 +325,17 @@ abstract class FormWizardBase extends FormBase implements FormWizardInterface {
    */
   public function previous(array &$form, FormStateInterface $form_state) {
     $cached_values = $form_state->getTemporaryValue('wizard');
-    $form_state->setRedirect($this->getRouteName(), $this->getPreviousParameters($cached_values));
+    $prev_parameters = $this->getPreviousParameters($cached_values);
+    // Redirect for non ajax forms.
+    if (!$form_state->get('ajax')) {
+      $form_state->setRedirect($this->getRouteName(), $prev_parameters);
+    }
+    else {
+      // Switch step for ajax forms.
+      if (!empty($prev_parameters['step'])) {
+        $this->step = $prev_parameters['step'];
+      }
+    }
   }
 
   /**
