diff --git a/js/views_ajax_history.js b/js/views_ajax_history.js
index a6455c6..0bae8e8 100644
--- a/js/views_ajax_history.js
+++ b/js/views_ajax_history.js
@@ -235,11 +235,14 @@
    *   jQuery.ajax options.
    */
   Drupal.Ajax.prototype.beforeSend = function (xmlhttprequest, options) {
-    // Override the URL to not contain any fields that were submitted.
-    options.url = drupalSettings.views.ajax_path + '?' + Drupal.ajax.WRAPPER_FORMAT + '=drupal_ajax';
+    // Check that the Ajax call is coming from a View.
+    if(typeof options.url != 'undefined' && options.url.indexOf('/views/ajax') != -1) {
+      // Override the URL to not contain any fields that were submitted.
+      options.url = drupalSettings.views.ajax_path + '?' + Drupal.ajax.WRAPPER_FORMAT + '=drupal_ajax';
 
-    // Call the original Drupal method with the right context.
-    beforeSend.apply(this, arguments);
+      // Call the original Drupal method with the right context.
+      beforeSend.apply(this, arguments);
+    }
   }
 
 })(jQuery, Drupal, drupalSettings);
